clr-k8s-examples/setup_system.sh

```diff
@@ -3,85 +3,159 @@
+  fi
+}
 
-sudo systemctl daemon-reload
-# This will fail at this point, but puts it into a retry loop that
-# will therefore startup later once we have configured with kubeadm.
-echo "The following kubelet command may complain... it is not an error"
-sudo systemctl enable --now kubelet $RUNNER || true
+# daemon reload
+function daemon_reload() {
+  sudo systemctl daemon-reload
+}
+
+# enable kubelet for $RUNNER
+function enable_kubelet_runner() {
+  # This will fail at this point, but puts it into a retry loop that
+  # will therefore startup later once we have configured with kubeadm.
+  echo "The following kubelet command may complain... it is not an error"
+  sudo systemctl enable --now kubelet $RUNNER || true
+}
 
-#Ensure that the system is ready without requiring a reboot
-sudo swapoff -a
-sudo systemctl restart systemd-modules-load.service
+# ensure that the system is ready without requiring a reboot
+function ensure_system_ready() {
+  sudo swapoff -a
+  sudo systemctl restart systemd-modules-load.service
+}
 
+# add proxy if found
+function setup_proxy(){
 set +o nounset
 if [[ ${http_proxy} ]] || [[ ${HTTP_PROXY} ]]; then
-	echo "Setting up proxy stuff...."
-	# Setup IP for users too
-	sed_val=${ADD_NO_PROXY//\//\\/}
-	[ -f /etc/environment ] && sudo sed -i "/no_proxy/I s/$/,${sed_val}/g" /etc/environment
-	if [ -f /etc/profile.d/proxy.sh ]; then
-		sudo sed -i "/no_proxy/I s/\"$/,${sed_val}\"/g" /etc/profile.d/proxy.sh
-	else
-		echo "Warning, failed to find /etc/profile.d/proxy.sh to edit no_proxy line"
-	fi
-
-	services=($RUNNER 'kubelet')
-	for s in "${services[@]}"; do
-		sudo mkdir -p "/etc/systemd/system/${s}.service.d/"
-		cat <<EOF | sudo bash -c "cat > /etc/systemd/system/${s}.service.d/proxy.conf"
+  echo "Setting up proxy stuff...."
+  # Setup IP for users too
+  sed_val=${ADD_NO_PROXY//\//\\/}
+  [ -f /etc/environment ] && sudo sed -i "/no_proxy/I s/$/,${sed_val}/g" /etc/environment
+  if [ -f /etc/profile.d/proxy.sh ]; then
+    sudo sed -i "/no_proxy/I s/\"$/,${sed_val}\"/g" /etc/profile.d/proxy.sh
+  else
+    echo "Warning, failed to find /etc/profile.d/proxy.sh to edit no_proxy line"
+  fi
+
+  services=($RUNNER 'kubelet')
+  for s in "${services[@]}"; do
+    sudo mkdir -p "/etc/systemd/system/${s}.service.d/"
+    cat <<EOF | sudo bash -c "cat > /etc/systemd/system/${s}.service.d/proxy.conf"
 [Service]
 Environment="HTTP_PROXY=${http_proxy}"
 Environment="HTTPS_PROXY=${https_proxy}"
 Environment="SOCKS_PROXY=${socks_proxy}"
 Environment="NO_PROXY=${no_proxy},${ADD_NO_PROXY}"
 EOF
-	done
+  done
 fi
 set -o nounset
+}
+
+# init for performing any pre tasks
+function init(){
+  echo ""
+}
+
+###
+# main
+##
+set +o nounset
+if [[ -n "${CLRK8S_OS}" ]]; then
+  # shellcheck disable=SC1090
+  source "$(dirname "$0")/setup_system_${CLRK8S_OS}.sh"
+fi
+set -o nounset
+
+echo "Init..."
+init
+echo "Setting OS Version..."
+upate_os_version
+echo "Adding OS Dependencies..."
+add_os_deps
```

Comment:
If we are calling setup_system.sh from another OS, won't this step fail and stop the script?

Reply:
Still a WIP but this does work. I can invoke on Ubuntu with `CLRK8S_OS=ubuntu ./setup_system.sh` where the `setup_system_ubuntu.sh` contains 
```
function upate_os_version() {
  sudo apt-get update
}

function add_os_deps() {
  # general
  sudo apt-get install -y ca-certificates apt-transport-https curl
  ...
}

function init(){
  echo "Setup for Ubuntu..."
  RUNNER="crio"
}
```
The above functions overload the original ones and the first two MUST be overridden because the original counterparts contain `swupd` commands.  I am testing now and have it working on Ubuntu. This allows us to accommodate other OSes without keeping files in tree.